experimental/dds/tree2/src/class-tree/schemaCreationUtilities.ts

```diff
@@ -0,0 +1,144 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { UsageError } from "@fluidframework/telemetry-utils";
+import { fail } from "../util";
+import { EmptyObject } from "../feature-libraries";
+import { SchemaFactory } from "./schemaFactory";
+// This type is referred to by the base type generated by schema.
+// Adding the import of it here prevents TypeScript from generating inline includes for it in the d.ts file.
+// If we let TypeScript generate these includes, they use relative paths which break API extractor's rollup.
+// eslint-disable-next-line unused-imports/no-unused-imports
+import { type NodeKind as _dummy } from "./schemaTypes";
+
+/**
+ * Create a schema for a node with no state.
+ * @remarks
+ * This is commonly used in unions when the only information needed is which kind of node the value is.
+ * Enums are a common example of this pattern.
+ * @alpha
+ */
+export function singletonSchema<TScope extends string, TName extends string>(
+	factory: SchemaFactory<TScope>,
+	name: TName,
+) {
+	return class SingletonSchema extends factory.object(name, {}) {
+		public constructor(data?: EmptyObject) {
+			super(data ?? {});
+		}
+		public get value(): TName {
+			return name;
+		}
+	};
+}
+
+/**
+ * Converts an enum into a collection of schema which can be used in a union.
+ * @alpha
+ */
+export function adaptEnum<TScope extends string, const TEnum extends Record<string, string>>(
```

Comment:
This only works for string-based enums, but presumably it wouldn't be too hard to extend this to capture number-based enums, right?

Reply:
Added note to docs.